client/settings/fraud-protection/advanced-settings/index.tsx

```diff
@@ -154,9 +162,31 @@ const FraudProtectionAdvancedSettingsPage: React.FC = () => {
 			.every( Boolean );
 	};
 
+	const checkAnyRuleFilterEnabled = (
+		settings: ProtectionSettingsUI
+	): boolean => {
+		return Object.values( settings ).some( ( setting ) => setting.enabled );
+	};
+
 	const handleSaveSettings = () => {
 		if ( validateSettings( protectionSettingsUI ) ) {
-			if ( ProtectionLevel.ADVANCED !== currentProtectionLevel ) {
+			const settings = writeRuleset( protectionSettingsUI );
+
+			if ( ! checkAnyRuleFilterEnabled( protectionSettingsUI ) ) {
+				if ( ProtectionLevel.BASIC !== currentProtectionLevel ) {
+					updateProtectionLevel( ProtectionLevel.BASIC );
+					updateAdvancedFraudProtectionSettings( settings );
+					saveSettings();
+				}
+				dispatch( 'core/notices' ).createErrorNotice(
+					__(
+						'Protection level set to "basic", at least one risk filter needs to be enabled for advanced protection.',
+						'woocommerce-payments'
+					)
+				);
+
+				return;
```

Comment:
I think the `wcpaySettings?.accountStatus?.fraudProtection?.declineOnAVSFailure` bit below is important because it sets the `declineOnAVSFailure ` value on the global `wcpaySettings ` object. Because of that, it would be good to refactor the logic a bit to only use the early return when we don't want to perform any updates.

This is my suggestion – I also changed the order of the function to remove the `else` from the main `if` statement so that we can remove an `if` level.

```diff
diff --git a/client/settings/fraud-protection/advanced-settings/index.tsx b/client/settings/fraud-protection/advanced-settings/index.tsx
index 36ed53fb2..f18b70a62 100644
--- a/client/settings/fraud-protection/advanced-settings/index.tsx
+++ b/client/settings/fraud-protection/advanced-settings/index.tsx
@@ -169,55 +169,59 @@ const FraudProtectionAdvancedSettingsPage: React.FC = () => {
 	};
 
 	const handleSaveSettings = () => {
-		if ( validateSettings( protectionSettingsUI ) ) {
-			const settings = writeRuleset( protectionSettingsUI );
-
-			if ( ! checkAnyRuleFilterEnabled( protectionSettingsUI ) ) {
-				if ( ProtectionLevel.BASIC !== currentProtectionLevel ) {
-					updateProtectionLevel( ProtectionLevel.BASIC );
-					updateAdvancedFraudProtectionSettings( settings );
-					saveSettings();
-				}
+		if ( ! validateSettings( protectionSettingsUI ) ) {
+			window.scrollTo( {
+				top: 0,
+			} );
+			return;
+		}
+
+		if ( ! checkAnyRuleFilterEnabled( protectionSettingsUI ) ) {
+			if ( ProtectionLevel.BASIC === currentProtectionLevel ) {
 				dispatch( 'core/notices' ).createErrorNotice(
 					__(
-						'Protection level set to "basic", at least one risk filter needs to be enabled for advanced protection.',
+						'At least one risk filter needs to be enabled for advanced protection.',
 						'woocommerce-payments'
 					)
 				);
-
 				return;
-			} else if ( ProtectionLevel.ADVANCED !== currentProtectionLevel ) {
-				updateProtectionLevel( ProtectionLevel.ADVANCED );
-				dispatch( 'core/notices' ).createSuccessNotice(
-					__(
-						'Current protection level is set to "advanced".',
-						'woocommerce-payments'
-					)
-				);
-			}
-
-			// Persist the AVS verification setting until the account cache is updated locally.
-			if (
-				wcpaySettings?.accountStatus?.fraudProtection
-					?.declineOnAVSFailure
-			) {
-				wcpaySettings.accountStatus.fraudProtection.declineOnAVSFailure = settings.some(
-					( setting ) => setting.key === 'avs_verification'
-				);
 			}
 
-			updateAdvancedFraudProtectionSettings( settings );
+			updateProtectionLevel( ProtectionLevel.BASIC );
+			dispatch( 'core/notices' ).createErrorNotice(
+				__(
+					'Current protection level is set to "basic". At least one risk filter needs to be enabled for advanced protection.',
+					'woocommerce-payments'
+				)
+			);
+		} else if ( ProtectionLevel.ADVANCED !== currentProtectionLevel ) {
+			updateProtectionLevel( ProtectionLevel.ADVANCED );
+			dispatch( 'core/notices' ).createSuccessNotice(
+				__(
+					'Current protection level is set to "advanced".',
+					'woocommerce-payments'
+				)
+			);
+		}
 
-			saveSettings();
+		const settings = writeRuleset( protectionSettingsUI );
 
-			recordEvent( 'wcpay_fraud_protection_advanced_settings_saved', {
-				settings: JSON.stringify( settings ),
-			} );
-		} else {
-			window.scrollTo( {
-				top: 0,
-			} );
+		// Persist the AVS verification setting until the account cache is updated locally.
+		if (
+			wcpaySettings?.accountStatus?.fraudProtection?.declineOnAVSFailure
+		) {
+			wcpaySettings.accountStatus.fraudProtection.declineOnAVSFailure = settings.some(
+				( setting ) => setting.key === 'avs_verification'
+			);
 		}
+
+		updateAdvancedFraudProtectionSettings( settings );
+
+		saveSettings();
+
+		recordEvent( 'wcpay_fraud_protection_advanced_settings_saved', {
+			settings: JSON.stringify( settings ),
+		} );
 	};
 
 	// Hack to make "Payments > Settings" the active selected menu item.

```

Let me know what you think about it!

Reply:
Thanks a lot, Eduardo, the changes make it much better structured, I overcomplicated it 😵

Merged https://github.com/Automattic/woocommerce-payments/pull/8762/commits/9469a2000ccafe1b52bf488bb793fce8da246ca6